File: macos/Sources/GlyphstrokeCore/StandardActions.swift
import Foundation

/// Стандартные действия системы: «копировать», «свернуть окно» и прочие.
///
/// Зачем отдельный тип действия, когда есть «клавиши». Во-первых, человеку не
/// нужно помнить, что «вставить» — это ⌘V, а «назад» в браузере — ⌘[; он
/// выбирает из списка. Во-вторых и главное, запись остаётся верной при
/// переносе: файл жеста со «стандартным» действием работает и на маке, и на
/// Linux, потому что там то же действие — `ctrl+v` и `alt+Left`.
///
/// **Имена обязаны совпадать с таблицей в `glyphstroke/standard.py`** версии для
/// Linux. Разъедутся — перенесённый жест молча перестанет работать, а это ровно
/// та беда, ради которой тип и заводился. Проверка совпадения — в тестах.
public enum StandardActions {
    /// Что делать: нажать сочетание или выполнить действие над окном.
    public enum Kind: String {
        case keys
        case window
    }

    public struct Entry {
        public let id: String
        /// Название для списка в редакторе.
        public let title: String
        public let kind: Kind
        public let value: String
    }

    /// Порядок здесь смысловой, а не алфавитный: в списке правка идёт рядом с
    /// правкой, окна рядом с окнами.
    public static let all: [Entry] = [
        // правка
        Entry(id: "copy", title: "Копировать", kind: .keys, value: "cmd+c"),
        Entry(id: "cut", title: "Вырезать", kind: .keys, value: "cmd+x"),
        Entry(id: "paste", title: "Вставить", kind: .keys, value: "cmd+v"),
        // На маке «удалить» — это клавиша Delete, которая на самом деле забой;
        // forward delete там жмут через fn и в обычной работе не используют.
        Entry(id: "delete", title: "Удалить", kind: .keys, value: "backspace"),
        Entry(id: "undo", title: "Отменить", kind: .keys, value: "cmd+z"),
        Entry(id: "redo", title: "Повторить", kind: .keys, value: "cmd+shift+z"),
        Entry(id: "select-all", title: "Выделить всё", kind: .keys, value: "cmd+a"),
        // файлы
        Entry(id: "new", title: "Создать", kind: .keys, value: "cmd+n"),
        Entry(id: "open", title: "Открыть", kind: .keys, value: "cmd+o"),
        Entry(id: "save", title: "Сохранить", kind: .keys, value: "cmd+s"),
        Entry(id: "print", title: "Печать", kind: .keys, value: "cmd+p"),
        Entry(id: "find", title: "Найти", kind: .keys, value: "cmd+f"),
        // вкладки и страницы
        Entry(id: "new-tab", title: "Новая вкладка", kind: .keys, value: "cmd+t"),
        Entry(id: "close-tab", title: "Закрыть вкладку", kind: .keys, value: "cmd+w"),
        Entry(id: "next-tab", title: "Следующая вкладка", kind: .keys, value: "cmd+shift+bracketright"),
        Entry(id: "prev-tab", title: "Предыдущая вкладка", kind: .keys, value: "cmd+shift+bracketleft"),
        Entry(id: "back", title: "Назад", kind: .keys, value: "cmd+bracketleft"),
        Entry(id: "forward", title: "Вперёд", kind: .keys, value: "cmd+bracketright"),
        Entry(id: "refresh", title: "Обновить", kind: .keys, value: "cmd+r"),
        // масштаб
        Entry(id: "zoom-in", title: "Крупнее", kind: .keys, value: "cmd+equal"),
        Entry(id: "zoom-out", title: "Мельче", kind: .keys, value: "cmd+minus"),
        Entry(id: "zoom-reset", title: "Обычный размер", kind: .keys, value: "cmd+0"),
        // окна
        Entry(id: "minimize", title: "Свернуть окно", kind: .window, value: "minimize"),
        // «Развернуть» на маке — это полноэкранный режим: отдельной кнопки
        // «максимизировать», как в Windows и GNOME, здесь нет.
        Entry(id: "maximize", title: "Развернуть окно", kind: .window, value: "fullscreen"),
        Entry(id: "unmaximize", title: "Вернуть размер окна", kind: .window, value: "unfullscreen"),
        Entry(id: "fullscreen", title: "Во весь экран", kind: .window, value: "fullscreen"),
        Entry(id: "close-window", title: "Закрыть окно", kind: .window, value: "close"),
        Entry(id: "switch-window", title: "Переключить окно", kind: .keys, value: "cmd+tab"),
        // прочее
        Entry(id: "quit", title: "Завершить программу", kind: .keys, value: "cmd+q"),
        Entry(id: "screenshot", title: "Снимок экрана", kind: .keys, value: "cmd+shift+4"),
    ]

    private static let byID: [String: Entry] = Dictionary(
        uniqueKeysWithValues: all.map { ($0.id, $0) })

    public static func entry(_ id: String) -> Entry? {
        byID[id.trimmingCharacters(in: .whitespaces).lowercased()]
    }

    /// Название действия для списков и журнала.
    public static func title(_ id: String) -> String {
        entry(id)?.title ?? id
    }

    /// Название вместе с сочетанием: «Вставить — ⌘V».
    ///
    /// Сочетание показывается прямо в списке: выбирая пункт из трёх десятков,
    /// человек обычно хочет знать, что именно нажмётся. У действий над окном
    /// показывать нечего — там не клавиши, а прямая команда окну.
    public static func describe(_ id: String) -> String {
        guard let entry = entry(id) else { return id }
        guard entry.kind == .keys else { return entry.title }
        return "\(entry.title) — \(symbols(entry.value))"
    }

    /// `cmd+shift+bracketright` → `⌘⇧]`.
    ///
    /// На маке сочетания принято писать значками, и запись вида «cmd+shift+]»
    /// в родном интерфейсе смотрится чужеродно.
    public static func symbols(_ combination: String) -> String {
        let names: [String: String] = [
            "cmd": "⌘", "command": "⌘", "super": "⌘", "meta": "⌘", "win": "⌘",
            "ctrl": "⌃", "control": "⌃",
            "alt": "⌥", "option": "⌥", "opt": "⌥",
            "shift": "⇧", "fn": "fn",
            "bracketleft": "[", "bracketright": "]",
            "equal": "=", "minus": "−", "grave": "`", "comma": ",", "period": ".",
            "slash": "/", "backslash": "\\", "semicolon": ";", "apostrophe": "'",
            "backspace": "⌫", "delete": "⌫", "forwarddelete": "⌦",
            "tab": "⇥", "space": "␣", "return": "↩", "enter": "↩", "escape": "⎋", "esc": "⎋",
            "left": "←", "right": "→", "up": "↑", "down": "↓",
            "page_up": "⇞", "pageup": "⇞", "page_down": "⇟", "pagedown": "⇟",
            "home": "↖", "end": "↘",
        ]
        var modifiers = ""
        var key = ""
        for part in combination.split(separator: "+") {
            let name = part.trimmingCharacters(in: .whitespaces).lowercased()
            let symbol = names[name] ?? name.uppercased()
            if ["⌘", "⌃", "⌥", "⇧", "fn"].contains(symbol) {
                modifiers += symbol
            } else {
                key = symbol
            }
        }
        return modifiers + key
    }
}

File: windows/src/Glyphstroke.Core/StandardActions.cs
namespace Glyphstroke.Core;

/// <summary>Что делать по стандартному действию: клавиши или команда окну.</summary>
public enum StandardKind
{
    Keys,
    Window,
}

public sealed record StandardAction(string Id, string Title, StandardKind Kind, string Value);

/// <summary>
/// Стандартные действия системы: «копировать», «свернуть окно» и прочие.
/// </summary>
/// <remarks>
/// Зачем отдельный тип действия, когда есть «клавиши». Во-первых, человеку не
/// нужно помнить сочетания — он выбирает из списка. Во-вторых и главное, запись
/// остаётся верной при переносе: файл жеста со «стандартным» действием работает
/// на всех трёх системах, потому что клавиши подставляет та, где он выполняется.
///
/// <b>Имена обязаны совпадать</b> с таблицами в <c>glyphstroke/standard.py</c>
/// (Linux) и <c>StandardActions.swift</c> (macOS). Разъедутся — перенесённый
/// жест молча перестанет работать, а это ровно та беда, ради которой тип и
/// заводился. Совпадение стережёт тест.
/// </remarks>
public static class StandardActions
{
    /// <summary>Порядок смысловой, а не алфавитный: правка рядом с правкой.</summary>
    public static readonly IReadOnlyList<StandardAction> All = new List<StandardAction>
    {
        // правка
        new("copy", "Копировать", StandardKind.Keys, "ctrl+c"),
        new("cut", "Вырезать", StandardKind.Keys, "ctrl+x"),
        new("paste", "Вставить", StandardKind.Keys, "ctrl+v"),
        new("delete", "Удалить", StandardKind.Keys, "delete"),
        new("undo", "Отменить", StandardKind.Keys, "ctrl+z"),
        new("redo", "Повторить", StandardKind.Keys, "ctrl+y"),
        new("select-all", "Выделить всё", StandardKind.Keys, "ctrl+a"),
        // файлы
        new("new", "Создать", StandardKind.Keys, "ctrl+n"),
        new("open", "Открыть", StandardKind.Keys, "ctrl+o"),
        new("save", "Сохранить", StandardKind.Keys, "ctrl+s"),
        new("print", "Печать", StandardKind.Keys, "ctrl+p"),
        new("find", "Найти", StandardKind.Keys, "ctrl+f"),
        // вкладки и страницы
        new("new-tab", "Новая вкладка", StandardKind.Keys, "ctrl+t"),
        new("close-tab", "Закрыть вкладку", StandardKind.Keys, "ctrl+w"),
        new("next-tab", "Следующая вкладка", StandardKind.Keys, "ctrl+pagedown"),
        new("prev-tab", "Предыдущая вкладка", StandardKind.Keys, "ctrl+pageup"),
        new("back", "Назад", StandardKind.Keys, "alt+left"),
        new("forward", "Вперёд", StandardKind.Keys, "alt+right"),
        new("refresh", "Обновить", StandardKind.Keys, "f5"),
        // масштаб
        new("zoom-in", "Крупнее", StandardKind.Keys, "ctrl+plus"),
        new("zoom-out", "Мельче", StandardKind.Keys, "ctrl+minus"),
        new("zoom-reset", "Обычный размер", StandardKind.Keys, "ctrl+0"),
        // окна
        new("minimize", "Свернуть окно", StandardKind.Window, "minimize"),
        new("maximize", "Развернуть окно", StandardKind.Window, "maximize"),
        new("unmaximize", "Вернуть размер окна", StandardKind.Window, "unmaximize"),
        new("fullscreen", "Во весь экран", StandardKind.Keys, "f11"),
        new("close-window", "Закрыть окно", StandardKind.Window, "close"),
        new("switch-window", "Переключить окно", StandardKind.Keys, "alt+tab"),
        // прочее
        new("quit", "Завершить программу", StandardKind.Keys, "alt+f4"),
        // Win+Shift+S — снимок области, штатный способ в Windows 10 и новее.
        new("screenshot", "Снимок экрана", StandardKind.Keys, "win+shift+s"),
    };

    private static readonly Dictionary<string, StandardAction> ById =
        All.ToDictionary(action => action.Id, StringComparer.OrdinalIgnoreCase);

    public static StandardAction? Find(string? id) =>
        id is not null && ById.TryGetValue(id.Trim(), out var action) ? action : null;

    /// <summary>Название действия для списков и журнала.</summary>
    public static string Title(string id) => Find(id)?.Title ?? id;

    /// <summary>Название вместе с сочетанием: «Вставить — Ctrl+V».</summary>
    /// <remarks>
    /// Сочетание показывается прямо в списке: выбирая пункт из трёх десятков,
    /// человек обычно хочет знать, что именно нажмётся. У действий над окном
    /// показывать нечего — там не клавиши, а прямая команда окну.
    /// </remarks>
    public static string Describe(string id)
    {
        var action = Find(id);
        if (action is null)
        {
            return id;
        }
        return action.Kind == StandardKind.Keys ? $"{action.Title} — {Pretty(action.Value)}" : action.Title;
    }

    /// <summary><c>ctrl+pagedown</c> → <c>Ctrl+PageDown</c>: так пишут в Windows.</summary>
    public static string Pretty(string combination)
    {
        var names = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
        {
            ["ctrl"] = "Ctrl", ["control"] = "Ctrl",
            ["alt"] = "Alt", ["shift"] = "Shift",
            ["win"] = "Win", ["super"] = "Win", ["meta"] = "Win", ["cmd"] = "Win",
            ["pageup"] = "PageUp", ["page_up"] = "PageUp",
            ["pagedown"] = "PageDown", ["page_down"] = "PageDown",
            ["left"] = "←", ["right"] = "→", ["up"] = "↑", ["down"] = "↓",
            ["plus"] = "+", ["minus"] = "−", ["equal"] = "=",
            ["delete"] = "Delete", ["backspace"] = "Backspace", ["tab"] = "Tab",
            ["space"] = "Space", ["enter"] = "Enter", ["return"] = "Enter", ["escape"] = "Esc",
        };
        var parts = combination.Split('+', StringSplitOptions.RemoveEmptyEntries)
            .Select(part => part.Trim())
            .Select(part => names.TryGetValue(part, out var pretty) ? pretty : part.ToUpperInvariant());
        return string.Join("+", parts);
    }
}
